### ISSUE TITLE
Trying to get started with this app

### ISSUE THREAD

**Issue by gdneye** [maximumstock/poe-stash-indexer#10]
Hey Maximumstock,

I've been looking around for an indexer and ran across this project of yours.  I'm not a programmer, but I am reasonably tech savvy, I've downloaded Docker and by following these instructions (http://localhost/tutorial/#the-command-you-just-ran) I was able to get their example Docker file working, but I haven't been able to figure out what I need to change from their example in order to get yours working.

I think maybe I'm having trouble building the initial container image.  If I try to use the command they gave (docker run -d -p 80:80 docker/getting-started), but replace the "getting-started" with "poe-stash-indexer" or something else I'm getting an error that it's unable to find image 'docker/poe-stash-indexer:latest' locally.

On the other hand if I try to use the "getting-started" image (but replacing everything else with your code) I can get all the way to the end of the demo, the container image seems to be working alright, but then the program seems to end or exit immediately.

Anyways, I'll try to keep looking, but I was wondering if you had steps to getting this program running.

Thanks!
Gdneye

**Comment by maximumstock** [comment]
Hey there,

thanks for reaching out! As quick disclaimer, this project isn't really feature-complete, so I'm not sure if its suitable for your needs. However, let's solve your problem first, shall we.

> I think maybe I'm having trouble building the initial container image. If I try to use the command they gave (docker run -d -p 80:80 docker/getting-started), but replace the "getting-started" with "poe-stash-indexer" or something else I'm getting an error that it's unable to find image 'docker/poe-stash-indexer:latest' locally.

Yep, that's probably because that requires my container image to be publicly available online in some kind of Docker registry. Since I only use Docker (or more specifically docker-compose) to develop and run this project on my personal server, I haven't published any images yet.

So, in order for you to run this indexer, you have to use [docker-compose](https://docs.docker.com/compose/). If you've cloned this project and `cd` into it, you can then run `docker-compose up` to start one container for the indexer and another one with a PostgreSQL database to hold the indexed data. 

You can also configure another standalone database outside of any Docker containers if that's what you want to do. If you plan to export some `.json` file (as the `README` currently states), then I shamefully have to inform you that this is not the current state of affairs. File export is currently not implemented :P.

**Comment by gdneye** [comment]
Hey man, thanks for responding.

First off, I think that this program will likely be enough for what I need.  All I'm really looking for is a program that does the following:

1) listening to the trade API
2) Parsing/Translating the API JSON data
3) Saving that data in a database
Bonus #4) If I can tell the program to start parsing at a certain public stash tab ID that would be nice, even if it's just a line in the code that I can change to whatever I need.

So I think your program does those things and I really don't need it to do anything else.  Once it's in a database I can use SQL and Excel to narrow down to a list of useful items and run any analysis I want to on those given items.

Second,

I'm still struggling a bit to get the program running.  I've tried to learn a bit more about images and containers, but I suspect that my issue is stemming from the fact that I don't fully understand them.  Here are the steps I've taken and where I seem to be running into an issue.

1) I appear to have installed Docker (and docker-compose) properly on my machine (Windows 10).  I also think I got WSL2 installed properly.
2) When I cd to the directory where this program is installed and run docker-compose up it appears as if the docker-compose and docker files run properly.  I get all the way to step 12/12 without any error messages.
3) After step 12/12 the program attempts to run, but is having an issue communicating with the server. 

Here is the log from the command prompt, and a screenshot of my images and containers.

https://imgur.com/a/qhWbdgI

I have a few guesses as to what the issue might be, but I don't know how to verify which one would be the problem, or how to fix them.

1) You mentioned in your previous post that your container image isn't publicly available so I need to create my own.  This may be my issue.  I am not sure if docker-compose up creates that container for me, or if I need to do something else to create it.

2) Alternatively, maybe the issue is that I haven't installed PostgreSQL on my machine?

3) My third thought is that maybe it's a file-system issue.  Since I'm using a Windows machine, maybe the program is looking for things in the wrong place.  I have installed and run Linux before, just for fun.  If you think I would have better luck installing Linux and attempting to run the program from Linux I'm happy to give that a shot.

I hope this is just a simple issue that I'm missing.  If it's more complicated I don't expect you to spend time solving my problems, but I wanted to check and was hoping that the issue is easily solveable.  Thanks in advance if you can help!

**Comment by maximumstock** [comment]
Hey there, 

let me try to go over your points.

> First off, I think that this program will likely be enough for what I need. All I'm really looking for is a program that does the following:
>
> 1. listening to the trade API
> 2. Parsing/Translating the API JSON data
> 3. Saving that data in a database
Bonus #4) If I can tell the program to start parsing at a certain public stash tab ID that would be nice, even if it's just a line in the code that I can change to whatever I need.
>
> So I think your program does those things and I really don't need it to do anything else. Once it's in a database I can use SQL and Excel to narrow down to a list of useful items and run any analysis I want to on those given items.

As of now, the indexer only processes currency and bulk items (eg. maps, essences, oils, etc.), but no uniques or rare items. Plus, there is a fixed schema in which the data is saved, which you can find [here](https://github.com/maximumstock/poe-stash-indexer/blob/master/migrations/2020-06-11-182409_offers/up.sql).

Processing ALL item data is not a problem, although I'd need to do some modifications, because this project was very much tailored to my personal needs. The fixed schema is not something I can make easily changeable or configurable.
Please make sure this is sufficient for you.

> Bonus #4) If I can tell the program to start parsing at a certain public stash tab ID that would be nice, even if it's just a line in the code that I can change to whatever I need.

I can add that, regardless.

>Second,
>
>I'm still struggling a bit to get the program running. I've tried to learn a bit more about images and containers, but I suspect that my issue is stemming from the fact that I don't fully understand them. Here are the steps I've taken and where I seem to be running into an issue.
>
>1. I appear to have installed Docker (and docker-compose) properly on my machine (Windows 10). I also think I got WSL2 installed properly.
>2. When I cd to the directory where this program is installed and run docker-compose up it appears as if the docker-compose and docker files run properly. I get all the way to step 12/12 without any error messages.
> 3. After step 12/12 the program attempts to run, but is having an issue communicating with the server.
Here is the log from the command prompt, and a screenshot of my images and containers.

I just checked and the issue you are facing is due to a bug in the `.env` file. Previously, the indexer tried to access the container housing the database via the name `poe`, which is false. The current master holds the correct database container name `db` and should work fine. I'm very sorry for the inconvenience!

>I have a few guesses as to what the issue might be, but I don't know how to verify which one would be the problem, or how to fix them.

Neither of below points are the actual issue, but I'll comment just FYI.

>1. You mentioned in your previous post that your container image isn't publicly available so I need to create my own. This may be my issue. I am not sure if docker-compose up creates that container for me, or if I need to do something else to create it.

If you look at `docker-compose.yml`, you can see a definition of a two container setup. The first one is a Postgres database, [referencing an official Docker image](https://github.com/maximumstock/poe-stash-indexer/blob/master/docker-compose.yaml#L4) (hosted on some publicly available server). The second container references the `Dockerfile` in [line 12](https://github.com/maximumstock/poe-stash-indexer/blob/master/docker-compose.yaml#L12). So instead of writing `.` and referencing a local `Dockerfile` you could also write something like `poe-stash-indexer` if such image would exist in your respective Docker registry, eg. DockerHub.

>2. Alternatively, maybe the issue is that I haven't installed PostgreSQL on my machine?

No this is fine, as the first container runs your PostgreSQL instance for you. 

Alternatively, you can of course run PostgreSQL inside your Windows user-space (and therefore outside of Docker) as well. Then however, you need to change the `.env` file so the `DATABASE_URL` points to `postgres://poe:poe@localhost/poe`. Also,
if you do so, you cannot run the indexer inside a container anymore, as AFAIK the containered indexer cannot easily access `localhost`, so it won't reach your locally running PostgreSQL instance. This is something I haven't found a proper solution to either.

Feel free to ask further questions.

**Comment by gdneye** [comment]
Thanks for the in-depth answers!

**> I just checked and the issue you are facing is due to a bug in the .env file. Previously, the indexer tried to access the container housing the database via the name poe, which is false.answers!**

This did appear to get me at least one step farther, but I'm running into a new error message.  Here is the new error message:

https://imgur.com/a/PWd9URM

Googling "exited with code 0" and "incomplete startup packet" both gave results indicating that those weren't problems, but the master-indexer container does not seem to be running properly.  

> As of now, the indexer only processes currency and bulk items (eg. maps, essences, oils, etc.), but no uniques or rare items. 

I see.  When I read the Readme file I was assuming the opposite.  When you stated that you "filtered out" all of the currency, etc items, I thought you meant that you filtered them out and removed them from the database.  I now understand that you meant that you filter them out, keep them, and remove the rest.

> Plus, there is a fixed schema in which the data is saved, which you can find here.

Upon reviewing the schema I can see why this project might not work for my needs.

Even though it's not what I was looking for I would still be interested in trying to get the program to run.  Half of the reason I am doing this project is to learn more about how to work with different programming languages and environments like Docker.  I feel like I've learned a few useful things in the last few days and I guess I dislike the idea of giving up when I'm this close to having the program running.  So if you're willing to help me resolve this new error message I'd appreciate it. But if not I totally understand.

**Comment by maximumstock** [comment]
Hey there, no problem! I'm glad I can help :)

> This did appear to get me at least one step farther, but I'm running into a new error message. Here is the new error message:
>
>https://imgur.com/a/PWd9URM
>
>Googling "exited with code 0" and "incomplete startup packet" both gave results indicating that those weren't problems, but the master-indexer container does not seem to be running properly.

Yep, I've encountered this as well, although not regularly. What seemingly helped is to rebuild the underlying Docker container via `docker-compose up --build indexer` and if that does not help, altering the source code of the indexer, ie. adding a `println!`statement in `src/main.rs` like in https://github.com/maximumstock/poe-stash-indexer/blob/master/src/main.rs#L25, then rebuilding the container again and then removing the additional change and rebuilding again. I know it sounds dumb, but it seems to invalidate some cache that makes the running indexer container able to talk to the database. 

Anyway, I don't expect you to use this workaround. I'll look into my `docker-compose.yml` setup and see where I made a mistake, since I'm sure its actually just a minor, annoying problem.

> I see. When I read the Readme file I was assuming the opposite. When you stated that you "filtered out" all of the currency, etc items, I thought you meant that you filtered them out and removed them from the database. I now understand that you meant that you filter them out, keep them, and remove the rest.

Good point, I made a commit to `README.me` to clarify that point.

>Upon reviewing the schema I can see why this project might not work for my needs.
>
>Even though it's not what I was looking for I would still be interested in trying to get the program to run. Half of the reason I am doing this project is to learn more about how to work with different programming languages and environments like Docker. I feel like I've learned a few useful things in the last few days and I guess I dislike the idea of giving up when I'm this close to having the program running. So if you're willing to help me resolve this new error message I'd appreciate it. But if not I totally understand.

No problem, we can make this work. The reason I structured the `Offer` data structure the way it is lies in my goal to use it for event sourcing, which benefits from really flat structures.

If you tell me what you expect in terms of schema, I can look into making this configurable. 
For example, if you plan to use Excel/Pandas then the `*.csv` exporter I'm currently writing in another branch might be of more interest to you.

In other words, just let me know.

**Comment by gdneye** [comment]
I'll be a bit busy this weekend, but thanks for the answers again.  I'll try out the workaround once I have a bit of time to mess around with this (likely sometime by Monday).  In terms of a schema, I think a flat structure probably isn't an issue for me.  I would honestly probably just do something like pull out the items and fields I need with a basic sql query and then export them to excel where I would do the bulk of the manipulation and analyzing.  (I would do most of the work in Excel because my skills in Excel are much better than SQL, not because it's necessarily a better program.  I just know 100% I can accomplish what I need to in Excel.)  So for example I might use SQL to pull all of the Watcher's Eyes and the fields I need, then export that to Excel to play around with until I find good patterns.  Then I'd build out some VBA or formulas to automate the process so I can just dump the SQL output into the Excel file and then the Excel file would automate getting me the data I needed from that file.

So having a flat structure isn't really an issue at all since I plan to export it to a flat database (Excel) anyways.  I think the issue is more that I would need more data fields on the items.  Item mods and values would be the main thing I think I need beyond what you are providing (definitely implicit and explicit, and maybe things like enchantments, synthesized, fractured, etc.  I'm not sure how things like fractured/synthesized/enchanted mods are presented in the API JSON so I'm not certain if that extra data would be needed or not.

I'm not sure what the advantages/disadvantages would be to the *.csv exporter but my plan IS to lean on my ability in Excel.  I would probably use the opportunity to try to improve my SQL and try to push myself a bit on my SQL abilities, and whenever I hit a wall with SQL or just get tired of learning transfer it into Excel and finish it up there, where I know I can do pretty much whatever I need to do.

But if you want to explain the pros/cons of the *.csv exporter I'd be interested in learning about it and I'm open to considering it or maybe just trying out both.

**Comment by maximumstock** [comment]
Okay, so understanding your use-case a little better, I'm fairly certain that the desired way of storage is a Postgres instance with optional JSONB columns to store nested data like mods on items.

Here a examples of the Stash Tab API representation of a Watcher's Eye and a synthesized rare ring:

```json
[
  {
    "verified": false,
    "w": 1,
    "h": 1,
    "icon": "https://web.poecdn.com/image/Art/2DItems/Jewels/ElderJewel.png?v=db0288d4c78a01e2e067b8a36c0bf1ee&w=1&h=1&scale=1",
    "league": "Heist",
    "id": "46f3960ae1430904d429368e699b3192c4d2eb7aa564a76282c65042e67ab5b7",
    "name": "Watcher's Eye",
    "typeLine": "Prismatic Jewel",
    "identified": true,
    "ilvl": 85,
    "note": "~price 20 chaos",
    "properties": [
      { "name": "Limited to", "values": [["1", 0]], "displayMode": 0 }
    ],
    "explicitMods": [
      "5% increased maximum Energy Shield",
      "6% increased maximum Life",
      "6% increased maximum Mana",
      "44% reduced Reflected Elemental Damage taken while affected by Purity of Elements",
      "56% increased Life Recovery from Flasks while affected by Vitality"
    ],
    "descrText": "Place into an allocated Jewel Socket on the Passive Skill Tree. Right click to remove from the Socket.",
    "flavourText": [
      "One by one, they stood their ground against a creature \r",
      "they had no hope of understanding, let alone defeating,\r",
      "and one by one, they became a part of it."
    ],
    "frameType": 3,
    "extended": { "category": "jewels", "baseType": "Prismatic Jewel" },
    "x": 22,
    "y": 4,
    "inventoryId": "Stash18"
  },
  {
    "verified": false,
    "w": 1,
    "h": 1,
    "icon": "https://web.poecdn.com/image/Art/2DItems/Rings/MoonstoneRing.png?v=4715c1b82b9843bf4b12f0aad25b3f26&w=1&h=1&scale=1&synthesised=1",
    "league": "Heist",
    "id": "35b8643fe62c45819691334a2f6fbe6f4c927fab3308a188273060f8eb3f4bf3",
    "synthesised": true,
    "name": "Dire Eye",
    "typeLine": "Synthesised Moonstone Ring",
    "identified": true,
    "ilvl": 83,
    "requirements": [
      { "name": "Level", "values": [["36", 0]], "displayMode": 0 }
    ],
    "implicitMods": [
      "12% increased Lightning Damage",
      "Curse Enemies with Frostbite on Hit, with 20% increased Effect"
    ],
    "explicitMods": [
      "+15 to Dexterity",
      "Adds 4 to 6 Physical Damage to Attacks",
      "Regenerate 4.3 Life per second",
      "+20% to Cold Resistance"
    ],
    "frameType": 2,
    "extended": {
      "category": "accessories",
      "subcategories": ["ring"],
      "prefixes": 1,
      "suffixes": 3,
      "baseType": "Moonstone Ring"
    },
    "x": 9,
    "y": 8,
    "inventoryId": "Stash77"
  }
]
```

There are several fields that are not present on other item types, eg. currency items.

So, assuming a Postgres setup with nested JSONB data in it, the indexer could run and store the data in a generic format, retaining all relevant information. You would then have to write custom SQL queries to produce the flat data output that your desired use-case requires.

> I'm not sure what the advantages/disadvantages would be to the *.csv exporter but my plan IS to lean on my ability in Excel. I would probably use the opportunity to try to improve my SQL and try to push myself a bit on my SQL abilities, and whenever I hit a wall with SQL or just get tired of learning transfer it into Excel and finish it up there, where I know I can do pretty much whatever I need to do.
>
>But if you want to explain the pros/cons of the *.csv exporter I'd be interested in learning about it and I'm open to considering it or maybe just trying out both.

Nevermind, actually, I don't think a general CSV exporter is useful to anyone at this stage. Well, maybe except to me, haha. The idea was just to have a single file to write CSV rows to, so you would save up on all the database overhead. But that is only relevant for very small amounts of data, as file sizes are pretty limited. So, database it is :)

**Comment by ek-01** [comment]
I'm having issues with the indexer saving to the database, I manage to get it all running, the indexer starts processing stashes.

```
Processed stash id "731542987-745511981-710982257-804782005-767589304"

Fetch: 652ms Deserialize: 236ms Parse: 0ms Persist: 0ms

Processed stash id "731543085-745516521-710984797-804786579-767592226"

Fetch: 836ms Deserialize: 223ms Parse: 0ms Persist: 0ms
```

But opening the database in dbeaver/datagrip shows that its not being fed any of the information.
```
PostgreSQL Database directory appears to contain a database; Skipping initialization


2020-10-27 06:56:50.831 UTC [1] LOG: starting PostgreSQL 13.0 (Debian 13.0-1.pgdg100+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 8.3.0-6) 8.3.0, 64-bit

2020-10-27 06:56:50.831 UTC [1] LOG: listening on IPv4 address "0.0.0.0", port 5432

2020-10-27 06:56:50.831 UTC [1] LOG: listening on IPv6 address "::", port 5432

2020-10-27 06:56:50.867 UTC [1] LOG: listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"

2020-10-27 06:56:50.891 UTC [25] LOG: database system was shut down at 2020-10-27 06:43:03 UTC

2020-10-27 06:56:50.902 UTC [1] LOG: database system is ready to accept connections

2020-10-27 06:56:51.296 UTC [32] LOG: incomplete startup packet

2020-10-27 06:57:12.839 UTC [36] LOG: invalid length of startup packet

2020-10-27 06:57:13.871 UTC [37] LOG: invalid length of startup packet

2020-10-27 06:57:14.452 UTC [38] LOG: invalid length of startup packet

2020-10-27 07:06:58.916 UTC [71] LOG: incomplete startup packet
```

**Comment by maximumstock** [comment]
@ek-01 Hello there, I've created a new issue for your problem to not mix concerns with the original issue. I'll continue commenting on that new issue.

**Comment by maximumstock** [comment]
Feel free to reopen this issue if I can keep helping you.
